packages/core/src/__tests__/state-manager.test.ts

```diff
@@ -71,75 +89,151 @@ describe('anchor', () => {
     const stateManager = ceramic.repository.stateManager
     const stream = await TileDocument.create(ceramic, INITIAL_CONTENT, null, { anchor: false })
     const stream$ = await ceramic.repository.load(stream.id, {})
 
+    await ceramic.repository.stateManager.anchor(stream$)
+    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.PENDING)
+    await waitForAnchorStatus(stream$, AnchorStatus.ANCHORED)
+
     const fakeHandleTip = jest.fn()
-    ;(stateManager as any)._handleTip = fakeHandleTip
-    // Handle tip needs to work once to create the commit when the CAS calls applyCommit with it
-    fakeHandleTip.mockImplementationOnce(realHandleTip)
-    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    const bound = fakeHandleTip.bind(stateManager)
+    ;(stateManager as any)._handleTip = bound
+
+    // Mock a throw as the first call
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+
+    // Mock the result of the original implementation as the second call - this one should return
+    // and stop the retrying mechanism
     fakeHandleTip.mockImplementationOnce(realHandleTip)
 
-    await new Promise<void>((resolve) => {
-      ceramic.repository.stateManager.anchor(stream$).add(resolve)
-    })
-    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.ANCHORED)
-    expect(fakeHandleTip).toHaveBeenCalledTimes(4)
+    // wait for 3 secs to let the retry mechanism work
+    await TestUtils.delay(3000)
+
+    // Check that fakeHandleTip was called only three times
+    expect(fakeHandleTip).toHaveBeenCalledTimes(2)
   })
 
-  test('anchor too many retries', async () => {
+  test(`_handleTip is retried up to three times, if it doesn't return`, async () => {
     const stateManager = ceramic.repository.stateManager
     const stream = await TileDocument.create(ceramic, INITIAL_CONTENT, null, { anchor: false })
     const stream$ = await ceramic.repository.load(stream.id, {})
 
+    await ceramic.repository.stateManager.anchor(stream$)
+    expect(stream$.value.anchorStatus).toEqual(AnchorStatus.PENDING)
+    await waitForAnchorStatus(stream$, AnchorStatus.ANCHORED)
+
     const fakeHandleTip = jest.fn()
+    fakeHandleTip.bind(stateManager)
     ;(stateManager as any)._handleTip = fakeHandleTip
+
+    // Mock a bunch of throws as consecutive calls to fakeHandleTip
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
     fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
-    fakeHandleTip.mockImplementationOnce(realHandleTip)
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+    fakeHandleTip.mockRejectedValueOnce(new Error('Handle tip failed'))
+
+    // wait for 3 secs to let the retry mechanism work
+    await TestUtils.delay(3000)
```

Comment:
Can we use `TestUtils.waitForState` for this to wait until the anchor fails, rather than using sleeping which is brittle?

Reply:
The way it works, as far as I can see (in current `develop`) is that whether `_handleTip(..)` is retried inside of `_handleAnchorCommit(...)` does not depend on the `anchorStatus` of the stream. The retries are made iff any error is thrown inside of this [block](https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/core/src/state-management/state-manager.ts#L274) , so effectively, retries are made when anything throws inside of `_handleTip(...)` or `publishTip(...)`.

Then it comes to the `anchorStatus`, at the time `_handleAnchorCommit(...)` is called the anchor had been already requested and in the case of this test, the commit had been successfully anchored.

So the only think this test tests is that, if  `_handleTip(...)` throws an error, it will be retried up to three times after `anchor(...)` is called.

Does that make sense, @stbrody ?

Reply:
Clarification: currently in `develop` `_handleAnchorCommit(...)` actually reverts the stream's anchor status to `FAILED` (check [here](https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/core/src/state-management/state-manager.ts#L305), but only if the anchored `tip` equals the current `state$.tip`, which is not the case during this test, because `state$.tip` actually points to the currently processed anchor commit.

Reply:
nope I'm still confused.  Nothing should even be calling `_handleTip` by this point in the test.  The anchor has already been successfully applied as of line 147.  I don't really understand why the assertion on line 165 isn't failing, given that there shouldn't be any anchor-related code running by the time the mocks are set up as far as I can tell?

Reply:
In my understanding, the assertion in line 165 is not failing, because the current behaviour of the system is such that `_handleAnchorCommit(...)` is called for a given tip **after** its corresponding state had been set to `ANCHORED`.

So if I'm not mistaken, I guess your confusion may be coming from the fact that it's not super clear what it currently means for a tip to be anchored.

Please, have a look at this test as it currently is on [develop](https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/core/src/__tests__/state-manager.test.ts#L105) - the behaviour is the same as in this PR. 

The goal of this PR is not to change the current behaviour, but rather just add a single async step before an anchor request is made.

Can we accept this as is (and maybe add a ticket to change the behaviour of anchor-state-tracking), or do you think I should continue digging into this in this PR, @stbrody :) ?